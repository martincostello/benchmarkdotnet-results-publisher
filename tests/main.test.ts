// Copyright (c) Martin Costello, 2024. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { ActionFixture } from './ActionFixture';
import { setup } from './fixtures';

import {
  afterAll,
  beforeAll,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';

const timeout = 45000;

describe('benchmarkdotnet-results-publisher', () => {
  beforeAll(() => {
    Date.now = jest.fn(() =>
      new Date(Date.UTC(2024, 7, 17, 12, 34, 56)).valueOf()
    );
  });

  describe('when publishing', () => {
    describe.each([
      ['new-benchmark', true],
      ['existing-benchmark', true],
      ['regression', false],
    ])('results for %s', (scenario: string, succeeds: boolean) => {
      let fixture: ActionFixture;

      beforeAll(async () => {
        await setup(scenario);
        fixture = new ActionFixture();

        await fixture.initialize(scenario);

        await fixture.run();
      }, timeout);

      afterAll(async () => {
        await fixture?.destroy();
      });

      test('reports the correct result', () => {
        expect(fixture.success).toEqual(succeeds);
      });

      test('does not log any errors', () => {
        expect(fixture.getErrors()).toEqual([]);
      });

      test('generates the expected GitHub step summary', async () => {
        expect(fixture.stepSummary).toMatchSnapshot();
      });
    });
  });
});
