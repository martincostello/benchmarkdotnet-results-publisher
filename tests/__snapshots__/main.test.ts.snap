// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`benchmarkdotnet-results-publisher when publishing results for existing-benchmark generates the expected GitHub step summary 1`] = `""`;

exports[`benchmarkdotnet-results-publisher when publishing results for new-benchmark generates the expected GitHub step summary 1`] = `
"<h3>MartinCostello.ProjectEuler.Benchmarks.PuzzleBenchmarks</h3>

\`\`\`

BenchmarkDotNet v0.14.0, Windows 11 (10.0.22621.4037/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 8.0.401
  [Host]   : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2
  ShortRun : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2

Job=ShortRun  IterationCount=3  LaunchCount=1  
WarmupCount=3  

\`\`\`
| Method | input | Mean                 | Error                | StdDev              | Gen0       | Gen1     | Gen2     | Allocated   |
|------- |------ |---------------------:|---------------------:|--------------------:|-----------:|---------:|---------:|------------:|
| **Solve**  | **001**   |         **2,500.867 ns** |         **6,305.292 ns** |         **345.6144 ns** |          **-** |        **-** |        **-** |        **24 B** |
| **Solve**  | **002**   |            **66.425 ns** |           **113.693 ns** |           **6.2319 ns** |     **0.0025** |        **-** |        **-** |        **24 B** |
| **Solve**  | **003**   |     **2,579,792.448 ns** |     **1,354,977.188 ns** |      **74,270.8722 ns** |          **-** |        **-** |        **-** |        **26 B** |
| **Solve**  | **004**   |    **31,424,122.917 ns** |   **107,366,841.583 ns** |   **5,885,138.9041 ns** |  **5500.0000** |  **62.5000** |        **-** |  **51923897 B** |
| **Solve**  | **005**   |   **273,567,950.000 ns** |   **131,983,433.248 ns** |   **7,234,457.3637 ns** |          **-** |        **-** |        **-** |       **424 B** |
| **Solve**  | **006**   |           **135.968 ns** |           **118.296 ns** |           **6.4842 ns** |     **0.0024** |        **-** |        **-** |        **24 B** |
| **Solve**  | **007**   |       **828,816.374 ns** |       **852,794.347 ns** |      **46,744.5360 ns** |    **25.3906** |        **-** |        **-** |    **240024 B** |
| **Solve**  | **008**   |       **479,759.375 ns** |       **344,094.905 ns** |      **18,861.0029 ns** |    **26.3672** |   **0.9766** |        **-** |    **251425 B** |
| **Solve**  | **009**   |       **810,745.117 ns** |       **376,971.069 ns** |      **20,663.0564 ns** |          **-** |        **-** |        **-** |        **48 B** |
| **Solve**  | **010**   |    **21,624,311.458 ns** |    **12,399,880.583 ns** |     **679,679.2990 ns** |   **218.7500** | **218.7500** | **218.7500** |   **2039727 B** |
| **Solve**  | **011**   |       **159,461.035 ns** |       **102,525.614 ns** |       **5,619.7749 ns** |    **14.6484** |   **3.6621** |        **-** |    **138192 B** |
| **Solve**  | **012**   |   **201,133,422.222 ns** |   **302,020,309.363 ns** |  **16,554,752.3451 ns** |          **-** |        **-** |        **-** |    **792157 B** |
| **Solve**  | **013**   |         **2,436.182 ns** |         **2,641.824 ns** |         **144.8073 ns** |     **0.5455** |        **-** |        **-** |      **5168 B** |
| **Solve**  | **014**   |    **26,101,823.958 ns** |    **13,744,584.573 ns** |     **753,387.0625 ns** |          **-** |        **-** |        **-** |        **36 B** |
| **Solve**  | **015**   |            **68.012 ns** |            **23.094 ns** |           **1.2658 ns** |     **0.0025** |        **-** |        **-** |        **24 B** |
| **Solve**  | **016**   |        **58,291.901 ns** |       **142,128.183 ns** |       **7,790.5253 ns** |     **3.2959** |        **-** |        **-** |     **31360 B** |
| **Solve**  | **017**   |     **2,807,034.961 ns** |       **957,393.756 ns** |      **52,477.9826 ns** |   **347.6563** |        **-** |        **-** |   **3273650 B** |
| **Solve**  | **018**   |         **7,867.365 ns** |        **11,942.512 ns** |         **654.6094 ns** |     **1.6327** |   **0.0229** |        **-** |     **15368 B** |
| **Solve**  | **019**   |        **25,168.776 ns** |        **31,429.336 ns** |       **1,722.7479 ns** |          **-** |        **-** |        **-** |        **24 B** |
| **Solve**  | **020**   |        **17,851.311 ns** |         **7,973.513 ns** |         **437.0552 ns** |     **1.8921** |        **-** |        **-** |     **18048 B** |
| **Solve**  | **021**   |       **111,815.922 ns** |       **296,161.093 ns** |      **16,233.5889 ns** |          **-** |        **-** |        **-** |       **824 B** |
| **Solve**  | **022**   |     **1,494,847.721 ns** |     **1,455,226.430 ns** |      **79,765.8714 ns** |    **87.8906** |  **58.5938** |  **29.2969** |    **831879 B** |
| **Solve**  | **023**   |    **34,444,868.750 ns** |    **33,529,270.832 ns** |   **1,837,852.4812 ns** |  **1562.5000** | **125.0000** |        **-** |  **15189745 B** |
| **Solve**  | **024**   |           **351.425 ns** |           **331.348 ns** |          **18.1623 ns** |     **0.0296** |        **-** |        **-** |       **280 B** |
| **Solve**  | **025**   |       **472,979.948 ns** |       **831,077.581 ns** |      **45,554.1667 ns** |   **119.1406** |        **-** |        **-** |   **1124321 B** |
| **Solve**  | **026**   |        **24,971.813 ns** |        **47,873.218 ns** |       **2,624.0926 ns** |     **7.6294** |        **-** |        **-** |     **71880 B** |
| **Solve**  | **027**   |    **14,762,249.479 ns** |    **11,570,579.639 ns** |     **634,222.5158 ns** |          **-** |        **-** |        **-** |        **30 B** |
| **Solve**  | **028**   |           **622.741 ns** |           **416.826 ns** |          **22.8476 ns** |     **0.0019** |        **-** |        **-** |        **24 B** |
| **Solve**  | **029**   |       **545,717.057 ns** |       **356,014.401 ns** |      **19,514.3507 ns** |    **49.8047** |  **49.8047** |  **49.8047** |    **202217 B** |
| **Solve**  | **030**   |    **58,749,975.000 ns** |     **4,346,407.607 ns** |     **238,241.2682 ns** |  **6250.0000** |        **-** |        **-** |  **58961882 B** |
| **Solve**  | **031**   |         **1,300.017 ns** |         **1,064.453 ns** |          **58.3463 ns** |     **0.0896** |        **-** |        **-** |       **856 B** |
| **Solve**  | **033**   |       **328,611.491 ns** |       **747,230.711 ns** |      **40,958.2369 ns** |   **123.5352** |        **-** |        **-** |   **1166544 B** |
| **Solve**  | **034**   |    **68,449,500.000 ns** |   **158,354,668.153 ns** |   **8,679,953.7405 ns** | **34500.0000** | **125.0000** |        **-** | **324590942 B** |
| **Solve**  | **035**   |    **22,671,433.333 ns** |     **6,921,744.025 ns** |     **379,404.1479 ns** |  **4406.2500** | **437.5000** | **218.7500** |  **40497720 B** |
| **Solve**  | **036**   |    **31,703,927.083 ns** |    **23,263,946.345 ns** |   **1,275,175.4050 ns** |  **8375.0000** |        **-** |        **-** |  **79347633 B** |
| **Solve**  | **037**   |    **18,000,764.583 ns** |    **11,187,356.356 ns** |     **613,216.7544 ns** |  **1718.7500** |        **-** |        **-** |  **16212364 B** |
| **Solve**  | **038**   |     **7,529,747.917 ns** |    **11,136,102.934 ns** |     **610,407.3814 ns** |  **2750.0000** | **125.0000** | **125.0000** |  **26378024 B** |
| **Solve**  | **040**   |    **85,195,033.333 ns** |    **21,872,493.699 ns** |   **1,198,905.1899 ns** | **18000.0000** | **500.0000** | **500.0000** | **171440343 B** |
| **Solve**  | **041**   |   **184,327,833.333 ns** |   **177,244,458.239 ns** |   **9,715,366.8800 ns** | **30000.0000** |        **-** |        **-** | **298478440 B** |
| **Solve**  | **042**   |       **124,043.127 ns** |        **65,353.531 ns** |       **3,582.2476 ns** |    **27.8320** |   **7.9346** |        **-** |    **262470 B** |
| **Solve**  | **044**   |             **8.245 ns** |             **2.185 ns** |           **0.1198 ns** |          **-** |        **-** |        **-** |           **-** |
| **Solve**  | **045**   |       **505,986.654 ns** |       **638,484.970 ns** |      **34,997.5158 ns** |          **-** |        **-** |        **-** |        **24 B** |
| **Solve**  | **046**   |        **70,599.434 ns** |        **56,679.260 ns** |       **3,106.7815 ns** |          **-** |        **-** |        **-** |        **24 B** |
| **Solve**  | **047**   |   **128,960,025.000 ns** |    **56,673,629.360 ns** |   **3,106,472.8744 ns** |  **1500.0000** |        **-** |        **-** |  **15633556 B** |
| **Solve**  | **048**   |    **10,403,805.469 ns** |     **5,062,807.272 ns** |     **277,509.5514 ns** |   **125.0000** |        **-** |        **-** |   **1225980 B** |
| **Solve**  | **049**   |    **28,575,333.333 ns** |    **27,026,936.815 ns** |   **1,481,437.6111 ns** |   **562.5000** |  **31.2500** |        **-** |   **5480972 B** |
| **Solve**  | **050**   |    **12,148,064.062 ns** |    **10,598,479.350 ns** |     **580,938.4185 ns** |   **968.7500** | **937.5000** | **937.5000** |   **6879904 B** |
| **Solve**  | **052**   |    **29,157,379.167 ns** |    **33,558,589.756 ns** |   **1,839,459.5503 ns** |  **8875.0000** |        **-** |        **-** |  **83543209 B** |
| **Solve**  | **054**   |     **8,433,740.885 ns** |    **19,121,506.594 ns** |   **1,048,114.3033 ns** |  **2000.0000** | **320.3125** |   **7.8125** |  **18880358 B** |
| **Solve**  | **055**   |    **13,043,958.333 ns** |    **18,024,416.412 ns** |     **987,979.0882 ns** |  **1765.6250** |        **-** |        **-** |  **16647582 B** |
| **Solve**  | **056**   |    **59,706,903.704 ns** |    **78,499,256.633 ns** |   **4,302,809.1574 ns** |  **5000.0000** |        **-** |        **-** |  **47579938 B** |
| **Solve**  | **059**   |     **5,180,591.927 ns** |     **8,273,027.300 ns** |     **453,472.5443 ns** |    **78.1250** |  **39.0625** |  **39.0625** |    **784969 B** |
| **Solve**  | **067**   |       **143,689.071 ns** |       **284,045.703 ns** |      **15,569.5035 ns** |    **34.4238** |   **3.4180** |        **-** |    **324343 B** |
| **Solve**  | **097**   | **5,029,947,233.333 ns** | **4,493,728,733.623 ns** | **246,316,436.2951 ns** |          **-** |        **-** |        **-** |  **12725704 B** |

"
`;

exports[`benchmarkdotnet-results-publisher when publishing results for regression generates the expected GitHub step summary 1`] = `""`;
